homeassistant/components/mqtt/config_flow.py

```diff
@@ -337,38 +450,151 @@ def validate_field(
         errors[field] = error
 
 
+@callback
+def _check_conditions(
+    platform_field: PlatformField, component: dict[str, Any] | None = None
+) -> bool:
+    """Only include field if one of conditions match, or no conditions are set."""
+    if platform_field.conditions is None or component is None:
+        return True
+    return any(
+        all(component.get(key) == value for key, value in condition.items())
+        for condition in platform_field.conditions
+    )
+
+
+@callback
+def calculate_merged_config(
+    merged_user_input: dict[str, Any],
+    data_schema_fields: dict[str, PlatformField],
+    component_data: dict[str, Any],
+) -> dict[str, Any]:
+    """Calculate merged config."""
+    base_schema_fields = {
+        key
+        for key, platform_field in data_schema_fields.items()
+        if _check_conditions(platform_field, component_data)
+    } - set(merged_user_input)
+    return {
+        key: value
+        for key, value in component_data.items()
+        if key not in base_schema_fields
+    } | merged_user_input
+
+
 @callback
 def validate_user_input(
     user_input: dict[str, Any],
     data_schema_fields: dict[str, PlatformField],
     errors: dict[str, str],
-) -> None:
+    component_data: dict[str, Any] | None,
+    config_validator: Callable[
+        [dict[str, Any], dict[str, str], list[str] | None], dict[str, str]
+    ]
+    | None = None,
+) -> dict[str, Any]:
     """Validate user input."""
-    for field, value in user_input.items():
+    # Merge sections
+    reset_fields: list[str] = []
+    merged_user_input: dict[str, Any] = {}
+    for key, value in user_input.items():
+        # Omit empty lists that are not allowed to be empty
+        if isinstance(value, dict):
+            merged_user_input.update(value)
+        else:
+            merged_user_input[key] = value
+
+    for field, value in merged_user_input.items():
         validator = data_schema_fields[field].validator
         try:
             validator(value)
         except (ValueError, vol.Invalid):
             errors[field] = data_schema_fields[field].error or "invalid_input"
 
+    if config_validator is not None:
+        if TYPE_CHECKING:
+            assert component_data is not None
+
+        config_validator(
+            calculate_merged_config(
+                merged_user_input, data_schema_fields, component_data
+            ),
+            errors,
+            reset_fields,
+        )
+
+    for field in reset_fields:
+        if component_data and field in component_data:
+            del component_data[field]
+
+    return merged_user_input
+
 
 @callback
 def data_schema_from_fields(
     data_schema_fields: dict[str, PlatformField],
     reconfig: bool,
+    component: dict[str, Any] | None = None,
```

Comment:
name this `component_config`, or `old_config` or `component_data` or something like that